src/mds/CDir.h

```diff
@@ -231,29 +240,44 @@ class CDir : public MDSCacheObject, public Counter<CDir> {
     inode->num_exporting_dirs--;
   }
 
-  version_t get_version() const { return fnode.version; }
-  void set_version(version_t v) { 
+  version_t get_version() const { return fnode->version; }
+  void update_projected_version() {
     ceph_assert(projected_fnode.empty());
-    projected_version = fnode.version = v; 
+    projected_version = fnode->version;
   }
   version_t get_projected_version() const { return projected_version; }
 
-  const fnode_t *get_projected_fnode() const {
-    if (projected_fnode.empty())
-      return &fnode;
-    else
-      return &projected_fnode.back();
+  void reset_fnode(fnode_const_ptr&& ptr) {
+    fnode = std::move(ptr);
+  }
+  void reset_fnode(const fnode_const_ptr& ptr) {
+    fnode = ptr;
+  }
+
+  const fnode_const_ptr& get_fnode() const {
+    return fnode;
+  }
+  fnode_t* _get_fnode() {
```

Comment:
These and the other evil accessors need comments explaining why we're breaking the `const`-ness of the ptr. The data members also need a comment. Otherwise the code misleads new developers.